util/topsql/reporter/datamodel.go

```diff
@@ -0,0 +1,673 @@
+		// stmtStats.DurationSum: []
+		//
+		// After:
+		//     tsIndex: [10000 => 0]
+		//     tsItems:
+		//             timestamp: [10000]
+		//             cpuTimeMs: [0]
+		//   stmtStats.ExecCount: [123]
+		// stmtStats.KvExecCount: [map{"1.1.1.1:1": 123}]
+		// stmtStats.DurationSum: [456]
+		//
+		newItem := newTsItem()
+		newItem.timestamp = timestamp
+		newItem.stmtStats = item
+		r.tsIndex[timestamp] = len(r.tsItems)
+		r.tsItems = append(r.tsItems, newItem)
+	}
+}
+
+// merge other record into r.
+// Attention, this function depend on r is sorted, and will sort `other` by timestamp.
+func (r *record) merge(other *record) {
+	if other == nil || len(other.tsItems) == 0 {
+		return
+	}
+	sort.Sort(other)
+	if len(r.tsItems) == 0 {
+		r.totalCPUTimeMs = other.totalCPUTimeMs
+		r.tsItems = other.tsItems
+		r.tsIndex = other.tsIndex
+		return
+	}
+	length := len(r.tsItems) + len(other.tsItems)
+	newTsItems := make(tsItems, 0, length)
+	i, j := 0, 0
+	for i < len(r.tsItems) && j < len(other.tsItems) {
+		if r.tsItems[i].timestamp == other.tsItems[j].timestamp {
+			newItem := newTsItem()
+			newItem.timestamp = r.tsItems[i].timestamp
+			newItem.cpuTimeMs = r.tsItems[i].cpuTimeMs + other.tsItems[j].cpuTimeMs
+			r.tsItems[i].stmtStats.Merge(&other.tsItems[j].stmtStats)
+			newItem.stmtStats = r.tsItems[i].stmtStats
+			newTsItems = append(newTsItems, newItem)
+			i++
+			j++
+		} else if r.tsItems[i].timestamp < other.tsItems[j].timestamp {
+			newItem := newTsItem()
+			newItem.timestamp = r.tsItems[i].timestamp
+			newItem.cpuTimeMs = r.tsItems[i].cpuTimeMs
+			newItem.stmtStats = r.tsItems[i].stmtStats
+			newTsItems = append(newTsItems, newItem)
+			i++
+		} else {
+			newItem := newTsItem()
+			newItem.timestamp = other.tsItems[j].timestamp
+			newItem.cpuTimeMs = other.tsItems[j].cpuTimeMs
+			newItem.stmtStats = other.tsItems[j].stmtStats
+			newTsItems = append(newTsItems, newItem)
+			j++
+		}
+	}
+	if i < len(r.tsItems) {
+		newTsItems = append(newTsItems, r.tsItems[i:]...)
+	}
+	if j < len(other.tsItems) {
+		newTsItems = append(newTsItems, other.tsItems[j:]...)
+	}
+	r.tsItems = newTsItems
+	r.totalCPUTimeMs += other.totalCPUTimeMs
+	r.rebuildTsIndex()
+}
+
+// rebuildTsIndex rebuilds the entire tsIndex based on tsItems.
+func (r *record) rebuildTsIndex() {
+	if len(r.tsItems) == 0 {
+		r.tsIndex = map[uint64]int{}
+		return
+	}
+	r.tsIndex = make(map[uint64]int, len(r.tsItems))
+	for index, item := range r.tsItems {
+		r.tsIndex[item.timestamp] = index
+	}
+}
+
+// toProto converts the record to the corresponding protobuf representation.
+func (r *record) toProto() tipb.TopSQLRecord {
+	capacity := len(r.tsItems)
+	if capacity == 0 {
+		return tipb.TopSQLRecord{
+			SqlDigest:  r.sqlDigest,
+			PlanDigest: r.planDigest,
+		}
+	}
+	timestampList := make([]uint64, 0, capacity)
+	cpuTimeList := make([]uint32, 0, capacity)
+	execCountList := make([]uint64, 0, capacity)
+	durationSumList := make([]uint64, 0, capacity)
```

Comment:
When a new metric comes, the `reporter/*` is still need to be modified. Maybe you can avoid it by refining the tipb model.

Reply:
I optimized the representation of the indicator list in `tipb`, from
```protobuf
message TopSQLRecord {
     bytes sql_digest = 1;
     bytes plan_digest = 2;
     repeated uint64 record_list_timestamp_sec = 10;
     repeated uint32 record_list_cpu_time_ms = 11;
     repeated uint64 record_list_stmt_exec_count = 12;
     repeated TopSQLStmtKvExecCount record_list_stmt_kv_exec_count = 13;
     repeated uint64 record_list_stmt_duration_sum_ns = 14;
}

message TopSQLStmtKvExecCount {
    map<string, uint64> exec_count = 1;
}
```
to
```protobuf
message TopSQLRecord {
     bytes sql_digest = 1;
     bytes plan_digest = 2;
     repeated TopSQLRecordItem items = 3;
}

message TopSQLRecordItem {
     uint64 timestamp_sec = 1;
     uint32 cpu_time_ms = 2;
     uint64 stmt_exec_count = 3;
     map<string, uint64> stmt_kv_exec_count = 4;
     uint64 stmt_duration_sum_ns = 5;
}
```

Then we can do protobuf conversion in a more fine-grained manner here:
```go
type tsItem struct {
	timestamp uint64
	cpuTimeMs uint32
	stmtStats stmtstats.StatementStatsItem
}

func (i *tsItem) toProto() *tipb.TopSQLRecordItem {
	return &tipb.TopSQLRecordItem{
		TimestampSec:    i.timestamp,
		CpuTimeMs:       i.cpuTimeMs,
		StmtExecCount:   i.stmtStats.ExecCount,
		StmtKvExecCount: i.stmtStats.KvStatsItem.KvExecCount,
		// Convert more indicators here.
	}
}

// All higher-order types' conversion below no longer need perception.

type tsItems []tsItem

func (ts tsItems) toProto() []*tipb.TopSQLRecordItem { /* we don't care */ }

type record struct { /* ... */ }

func (r *record) toProto() tipb.TopSQLRecord { /* we don't care */ }

type records []record

func (rs records) toProto() []tipb.TopSQLRecord { /* we don't care */ }
```

If a new indicator is added, we only need to add one line of code now.

Reply:
The only question now is whether the added line of code needs to be moved out of the `reporter` package (let `reporter` do not need any changes in future needs, even a single line of code). Since `tsItem` is a private type, moving out of the `reporter` package means that it needs to be made public, even though no package depends on `tsItem` except `reporter`.